File: c++/2110(공유기 설치)/2110.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int n;
int c;
int ans[200001];
int ret(0);

int main()
{
	cin >> n >> c;
	for(int i = 0; i < n; i++)
	{
		cin >> ans[i];
	}
	sort(ans,ans+n);
	int left = 0;
	int right = ans[n-1];

	while(left <= right)
	{
		int mid = (left + right) /2;
		int prev(0),cnt(1);
		for(int i = 1; i< n ;i++)
		{
			if(ans[i] - ans[prev] >= mid)
			{
				cnt++;
				prev = i;
			}
		}
		if (cnt >= c)
		{
			if(ret < mid)
				ret = mid;
			left = mid + 1;
		}
		else
			right = mid - 1;
	}
	cout << ret << '\n';
}

File: Swift/2110.swift
import Foundation

let input = readLine()!.split(separator: " ").map{ Int(String($0))!}
let n = input[0], c = input[1]

var house:[Int] = []

for _ in 0..<n{
    let num = Int(readLine()!)!
    house.append(num)
}

house = house.sorted()

var left = 1
var right = house.last! - house.first!

func isTrue(dist:Int) -> Bool{
    var cnt = 1
    var prev = 0
    for i in 1..<house.count{
        if house[i] - house[prev] >= dist{
            cnt += 1
            prev = i
        }
    }

    return cnt >= c
}

var maxDist = 1
while left <= right{
    let mid = (left + right) / 2
    if isTrue(dist: mid){
        maxDist = max(maxDist, mid)
        left = mid + 1
    }else{
        right = mid - 1
    }
}

print(maxDist)
